### PR TITLE
ui filter schema on new mssql connection widget

### PR DESCRIPTION
**Pull request by vcloarec**
mssql_schema_filter

### CODE REVIEW

**Review by wonder-sk — COMMENTED** [review]
Looking good, just one issue: the filtering of schemas should be done on the server side rather than getting everything and then filtering on the client side - that will also lower the amount of time needed to fetch all information... so in QgsMssqlConnectionItem::createChildren() we need to alter the SQL queries, e.g. when mUseGeometryColumns is true, and filtering enabled, the query would also contain something like `WHERE f_table_schema IN ('schema1', schema2')` - and something similar with mUserGeoemtryColumns=false and with mAllowGeometrylessTables=true

**Comment by vcloarec** [comment]
filtering is now on server side, and I have done some refactoring and allow this filtering from Data source manager widget

File: src/providers/mssql/qgsmssqlconnection.cpp [hunk 1]
```diff
@@ -468,6 +468,74 @@ QList<QgsVectorDataProvider::NativeType> QgsMssqlConnection::nativeTypes()
...
+      QVariantMap schemaSettingsForDatabase = schemaSettings.value( databaseName ).toMap();
+      //schema filter
+
+
+      QStringList schemaNames;
+      for ( const QString &schemaName : schemaSettingsForDatabase.keys() )
+      {
+        if ( schemaSettingsForDatabase.value( schemaName ).toBool() )
+        {
+          schemaNames.append( "'" + schemaName + "'" );
```
Comment (wonder-sk) [inline src/providers/mssql/qgsmssqlconnection.cpp]: We should use `QgsMssqlProvider::quotedValue( schemaName )` here instead of `"'" + schemaName + "'"` to make sure some characters are escaped properly (e.g. single quotes).

File: src/providers/mssql/qgsmssqldataitems.cpp [hunk 2]
```diff
@@ -270,24 +251,29 @@ QVector<QgsDataItem *> QgsMssqlConnectionItem::createChildren()
       }
     }
 
-    // add missing schemas (i.e., empty schemas)
-    const QString uri = connInfo();
-    const QStringList allSchemas = QgsMssqlConnection::schemas( uri, nullptr );
-    for ( const QString &schema : allSchemas )
+    if ( !mUseGeometryColumns )
```
Comment (wonder-sk) [inline src/providers/mssql/qgsmssqldataitems.cpp]: I am not sure what is the reason of this change - why this code would be run only when `if ( !mUseGeometryColumns )` ?

File: src/providers/mssql/qgsmssqldataitems.cpp [hunk 3]
```diff
@@ -270,24 +251,29 @@ QVector<QgsDataItem *> QgsMssqlConnectionItem::createChildren()
...
-        continue;
+      // add missing schemas (i.e., empty schemas)
+      const QString uri = connInfo();
+      const QStringList allSchemas = QgsMssqlConnection::schemas( uri, nullptr );
+      QVariantMap schemaSettings = mSchemaSettings.value( mDatabase ).toMap();
+      for ( const QString &schema : allSchemas )
+      {
+        if ( addedSchemas.contains( schema ) ||
+             ( schemaSettings.contains( schema ) && !schemaSettings.value( schema ).toBool() ) )
+          continue;
```
Comment (wonder-sk) [inline src/providers/mssql/qgsmssqldataitems.cpp]: This code for filtering should not be needed anymore if filtering is done on the server... or did I get something wrong here?

File: src/providers/mssql/qgsmssqldataitems.cpp (hunk 2 above)
Comment (vcloarec) [reply src/providers/mssql/qgsmssqldataitems.cpp]: this piece of code search in all the schemas for those that are not already added (that is the empty one), but if we we have "only use geometry columns" checked we don't want all but only those that are in the geometry columns, no?

File: src/providers/mssql/qgsmssqldataitems.cpp (hunk 3 above)
Comment (vcloarec) [reply src/providers/mssql/qgsmssqldataitems.cpp]: here, there is a new query that returns ALL the schema to search which are not already added. if we do not do another filtering the empty one not selected will appears in the browser.
It could be possible to filter, as well here, on the server side, but we have to modify `QgsMssqlConnection::schemas()` method

File: src/providers/mssql/qgsmssqldataitems.cpp (hunk 2 above)
Comment (wonder-sk) [reply src/providers/mssql/qgsmssqldataitems.cpp]: I think this should stay independent from the "use geometry columns" flag... if there is a schema which is empty, we still want to show it (even if there are no tables/views in it), so the user is not confused that it does not exist

File: src/providers/mssql/qgsmssqldataitems.cpp (hunk 3 above)
Comment (wonder-sk) [reply src/providers/mssql/qgsmssqldataitems.cpp]: Ah right, I just got confused...

May I suggest to split that `if` into two separate ifs - I always find it difficult to parse these combined ifs with multiple and, or, not clauses :-)

... and shouldn't we have the condition to use `mSchemasFilteringEnabled` instead of `schemaSettings.contains( schema )` ? With the latter, we could be showing a schema item for one which was recently added, which is not in the list of filtered schemas, but there would not be any tables listed with it...?

so my suggestion would be to change it to this:
```
if ( mSchemasFilteringEnabled && !schemaSettings.value( schema ).toBool() )
  continue;  // user does not want it to be shown

if ( addedSchemas.contains( schema ) )
  continue;
```
